File: ch3/3-5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void reverse(char s[]);
void itob(int n, char s[], int b);

int main(void)
{
    char s[100];
    long int n = 18237012837;

    itob(n, s, 16);
    printf("%s\n", s);
    return 0;
}

// reverse: reverse string s in place
void reverse(char s[])
{
    int c, i, j;
    for (i = 0, j = strlen(s)-1; i < j; i++, j--) {
        c = s[i];
        s[i] = s[j];
        s[j] = c;
    }
}

void itob(int n, char s[], int b)
{
    int i, sign, num;
    sign = n;
    n = abs(n); // sets n to be positive
    i = 0;
    do {
        // generate digits in reverse order
        num = n % b;
        if (num < 10) {
          s[i++] = num + '0';
        } else {
          s[i++] = num - 10 + 'a';
        }
    } while ((n /= b) > 0);

    if (sign < 0) {
        s[i++] = '-';
    }
    s[i] = '\0';
    reverse(s);
}

File: gkr/ch3/3-5.go
package main

import (
	"fmt"
	"math"
)

func main() {
	s1 := make([]byte, 100)
	s2 := make([]byte, 100)
	n := 530

	itob(n, s1, 16) // hex
	fmt.Printf("%d = %s\n", n, s1)
	itob(n, s2, 2) // base 2
	fmt.Printf("%d = %s\n", n, s2)
}

func itob(n int, s []byte, b int) {
	var i, sign, num int
	sign = n
	if sign < 0 {
		if n <= math.MinInt64 {
			n++
		}
		n = -n
	}
	for {
		num = n % b
		if num < 10 {
			s[i] = byte(num + '0')
			i++
		} else {
			s[i] = byte(num - 10 + 'a')
			i++
		}
		n = n / b
		if n == 0 {
			break
		}
	}
	if sign < 0 {
		s[i] = byte('-')
		i++
	}
	reverse(s, i)
}

func reverse(s []byte, l int) {
	var i, j int
	var c byte
	for j = l - 1; i < j; i, j = i+1, j-1 {
		c = s[i]
		s[i] = s[j]
		s[j] = c
	}
}
